### DISCUSSION TITLE
Node queue limit?

### DISCUSSION THREAD

**Discussion by SimonUnge** [rabbitmq/rabbitmq-server#10748] [Ideas]
Hi,

One of the tools we have  to make sure users do not overload the broker or cluster, which would lead to operational pain, is limits. 
One way to overload a broker is by creating too many queues, so I have been thinking about the possibility to introduce a 'queue limit' that is 'per node', or maybe 'per cluster' - there is already a queue limit per vhost. I would like to pick the community's brain a bit, as a per node limit for a node is not out of the box obvious - a queue is not really tied to a node, such as a connection or channel is, but instead with replication can live on more nodes, and even be moved around on nodes.

But, it would be useful to limit the queue numbers on a more global way than just per vhost - you can always bypass a limit by adding more vhosts etc. 

My 'simple' suggestion would be a *per node* queue limit - set in the config file -  that is only enforces when a queue is create 'externally', i.e via a client or via HTTP, but not 'internally', e.g the limit would not be taken into account during replication of queues from another node, if that node for some reason has a higher limit. (Preferably it would not interfere say the shovel plugin to declare queues either, even if limit is reached, but I am not sure it can be distinguished?)

The limit would count *all* queues on the local node, regardless if they are replicated, a leader or follower etc, as the would regardless take up resources and if the limit is reached, we want to stop the creation of new queues.

Technically it is fairly straight forward to add such a limit, but I want to hear concerns and input before any potential PR.

**Comment by SimonUnge** [comment]
Any thoughts @michaelklishin @lukebakken et al?

**Reply by lukebakken** [reply]
It's only been four days since you asked!

**Reply by SimonUnge** [reply]
haha true I am sorry for seeming pushy!

**Comment by mkuratczyk** [comment]
I agree that having such a limit would be very useful in some scenarios. Some questions that we'd need to answer:

1. How could we handle exclusive queues? They are always connection-local so you'd be able to declare them or not, based on where your connection happened to established.

2. Placement of classic non-mirrored queues would probably need to consider this limit. We need to reconsider this area anyway (https://github.com/rabbitmq/rabbitmq-server/issues/10702) and perhaps the new approach would need to consider node limits to avoid declaration failures when there's capacity on other nodes.

3. We would need to expose Prometheus metric(s) to allow calculating how much of the limit is already used. Existing `rabbitmq_queues` metric doesn't account for followers I believe so we would need either:
- the limit as a metric and a new counter for queues including followers (or for followers only)
- or a single new metric that exposes the percentage of node limit used

4. In a multi-tenant / multi-vhost environment, it could be too easy to exceed the node limit on some vhosts so that another vhost can't potentially declare a single queue. Probably worth considering a bunch of precautions / warnings. Otherwise we can expect a lot of "I can't declare a queue" issue reports. Perhaps a soft limit is also necessary to trigger a warning?

**Reply by SimonUnge** [reply]
Thank you Michal. I will do some thinking and digging and get back to the questions we need answered!

**Reply by SimonUnge** [reply]
Thoughts:
1. I would say for exclusive queues, we check the limit on the node the connection is at, and refuse it if over the limit. For classic non-mirrored queues, we could perhaps try to find another node in the cluster to add the queue to?
2. Regarding master_locator. ah, right,  some logic that tries to add the queue to a different node if the current one is "full".
3. Ill work in this last.
4. @mkuratczyk What is the RabbitMQ way of sending out precautions and warnings, other than say logs and via metrics? I agree that a configurable soft limit or percentage would be useful to add too!

**Reply by mkuratczyk** [reply]
1. I think the default/usual way of handling declaration errors would be to close the channel. But if the connection is with a node that has reached the limit, this will potentially just lead to a new channel being created within the same connection, another attempt and another failure. If that's indeed the case, I wonder if we should try to do something else here (close the connection to give it a chance to connect to a different node?)
4. I think we can follow the pattern for memory/disk alarms here - ultimately "running out queue capacity" is another way of running out of resources. We could mark the node as orange/red in the Management UI, log something, mention this in the `rabbitmqctl status` output, etc. In fact, perhaps we should introduce a new kind of alarm (a "limit alarm"?) and report this everywhere where alarms are listed?

**Reply by SimonUnge** [reply]
Ah, right!

I did notice that for limits on channels, once you reach the limit, we close the connection, so Ill see if I can do something similar for exclusive queues.

**Reply by SimonUnge** [reply]
I finally had some cycles to start looking at this again.

@mkuratczyk we discussed how to handle non mirrored classic queues, for example if the current node is at its limit for queues, we could declare the queue on another node. But, how about mirrored queues, or QQs or streams even. During declaration, if the node is full, would it make sense to see if another node has capacity for to be the leader, and somehow try to exclude the current node for the replication, or do we start treading on fairly complex waters by then? 
(or if things like x-queue-master-locator args is used)

**Reply by mkuratczyk** [reply]
I wouldn't even touch mirroring at this point. For QQs, with the `local` leader locator set, excluding the local node from the membership would be very strange...

**Reply by SimonUnge** [reply]
Yes I agree, it started to become hard to see a natural solution when it comes to mirroring.

**Reply by SimonUnge** [reply]
I'll get a PR up for review fairly soon, with only the hard limits for now, as I want to see that I am going in the right direction, especially with closing a connection from within a channel process.

**Reply by SimonUnge** [reply]
@mkuratczyk just to make sure I am on the same page as you are :)!
This is the current 'design' I am working on, in terms of the 'hard' limit and when a queue declaration gets rejected due to the limit.
(Metrics, soft limits etc I will work on later)

 #### Exclusive queues
Regardless of type (classic, classic mirror, QQ, Stream...), if the limit is reached on the node that receives the declaration request, the request is denied, and the CONNECTION is terminated.

#### Non Exclusive queues
For QQ, Stream and Classic MIRROR Queue: If the limit is reached on the node that receives the declaration request, the request is denied with the CHANNEL terminated.

For Classic Queue that is *NOT* mirrored:  If the limit is reached on the node that receives the declaration request, see if any other node in the cluster has capacity. If yes, redirect the declaration to that node. If no,  the request is denied with the CHANNEL terminated.


Is that what you had in mind too?

**Reply by mkuratczyk** [reply]
It sounds about right, but it's not like I have all of that figured out. I see you now created a PR, I guess we'll learn more as we play with it. Thanks!

**Reply by SimonUnge** [reply]
A rough version yes! But, kinda works but not super efficient way of counting queues per node etc, so will make some improvements and clean it up, but you are more than welcome to take a little look if you have time.

**Reply by SimonUnge** [reply]
@mkuratczyk OK, the PR is not in 'working state', with tests. I wonder if there is a more efficient way to count the queues per nodes though.

**Reply by mkuratczyk** [reply]
Having played with the PR and discussing this with other folks on the team, we have serious concerns about the feasibility of this feature. There are still many additional nueanced but critical "details" to work out - stream and QQ replica placement, queues explicitly declared with the `local` placement, AMQP 1.0 / MQTT / STOMP (what the behaviour should be for each) and on top of that, performance (at the very least - this should be zero-overhead if the limit is not set).

Can you please provide more detailed justification for this feature? Given that even a single queue can consume much more resources than a large number of idle / mostly idle queues, we feel like it's a lot of complexity for relatively little gain.

As you said, there's currently a per-vhost limit. We could additionally introduce a per-cluster limit for users with multiple vhosts. That would be a much simpler feature and perhaps sufficient for many of the same situations?

**Reply by SimonUnge** [reply]
Hey @mkuratczyk Yeah I agree the complexity of this feature grew exponentially, and there are many queue types and scenarios to take into account, and getting that right, will be a challenge. 

A much simpler 'per cluster' setting would absolutely be sufficient. Just want to make sure we mean the same thing. For me, a cluster wide limit would basically mean, as soon as the queue limit is hit, by looking at say `rabbit_db_queue:count/0`, no more queues can be created, even though in theory some node in the cluster have 'room'. Basically what the vhost limit does today, but looking at the entire cluster.

As a justification, it would be the same as for the vhost limit - avoid a broker becoming unresponsive (and in some cases crash), when the operator cannot control the actions of the client(s)/users. So, the same reasoning behind vhost limit, but acknowledging that having more vhost doesn't mean the cluster can handle more queues.

If I go ahead and make a much simpler cluster limit, where would such a limit live? I would suggest in the rabbit.conf file still, and you as an operator just have to make sure you configure your cluster correctly, with the same limits. Could of course do something more advanced and store the config in the database, or consolidate the config between the nodes (take the min value or something) but personally I would think its fine with setting it per node?

For now, I will close my current PR, and work on a simple version!

**Reply by michaelklishin** [reply]
Note that we do not have a perfect example of how such global limits would be stored.

For virtual hosts, they are a part of the virtual host metadata. For operator policies, they are, well, runtime parameters of a special kind, like all policies.

For per-node limits they are configuration settings.

I don't know if this would fit the global parameters store well or we'd need a yet another table/Khepri sub-tree.

**Reply by SimonUnge** [reply]
Right.. I'll get the logic for adhering to the limit in, and tinker with 'where' the limit would live and we can see what feels right!

**Reply by michaelklishin** [reply]
Thank you, @SimonUnge!
